input-output-hk/cardano-sl
# [CSL-2526] update changes after covering remaining fees

## Description

This pr tries to fix a bug on coin and fee selection, which was reported by QA.
The bug was caused because in some cases some changes were not included in the transaction, resulting in huge fees.

The CoinSelection works in many different stages:
- first given some payees, some inputs are selected which cover the expenses (fees are ignored for now). Changes are computed at this stage.
- then fees are computed and we reduce changes to cover the fees. If they are enough, we are done.
- if not new inputs are selected. 

The problem is that these additional inputs (last step) which cover the fees, may not be entirely consumed for the fees: we need to return what's left as change.

TODO:
- [x] Decide exactly how these changes will be returned: Do we create new additional change outputs or do we distribute them to the existing changes? If we decide the latter, do we need to equally distribute the additional change to all existing one, or just give everything to one of them? 
UPDATE: we split additional changes to existing change outputs. We do this in FromGeneric module (instead of the Generic or Fees) for 2 reason: in this module change outputs are flatten to a list (instead of list of lists) and also it is less polymorphic, which means we can directly use divide function from Core.
- [ ] Another issue is the additional fees that we may need to pay for these additional utxos and changes. This issue becomes too complex.
- [x] Add a sanity check after the coin selection, which reports an internal error if fees are not within a reasonable range.
- [ ] check `round` vs `ceiling` in the use of `calculateTxSizeLinear`.

## Linked issue

https://iohk.myjetbrains.com/youtrack/issue/CSL-2526

## QA Steps
Here are some steps to reproduce the bug with tome good probabilities:
- Create a wallet. 
- send 1 tx with amount > 1 ADA to it.
- send 2 txs with ammount in [100..1000] Lovelace to it.
- try to send < 100 Lovelace from it.

After trying the last step with different amounts and addresses, at some point the bug will appear and fees will be > 1 ADA.

The bug appears in these cases because the inputs in [100..1000] are enough to cover the <100 tx, but not enough to cover the fees. So an additional input needs to be selected. If we are "lucky" this input will be the > 1 ADA in which case we will have huge fees (without the fix).

## Type of change



## Screenshots (if available)

before
![bad](https://user-images.githubusercontent.com/11467473/47898973-c8275080-de80-11e8-89c6-26090193a0fc.png)

after
![good](https://user-images.githubusercontent.com/11467473/47898979-ce1d3180-de80-11e8-8027-e944955015ae.png)
